sdk/go/pulumi/provider.go

```diff
@@ -0,0 +1,271 @@
+		lastSep := strings.LastIndex(ref, "::")
+		if lastSep == -1 {
+			return nil, errors.Errorf("expected '::' in provider reference %s", ref)
+		}
+		urn := ref[0:lastSep]
+		id := ref[lastSep+2:]
+		providers[pkg] = newDependencyProviderResource(URN(urn), ID(id))
+	}
+	var parent Resource
+	if req.GetParent() != "" {
+		parent = newDependencyResource(URN(req.GetParent()))
+	}
+	opts := resourceOption(func(ro *resourceOptions) {
+		ro.Aliases = aliases
+		ro.DependsOn = dependencies
+		ro.Protect = req.GetProtect()
+		ro.Providers = providers
+		ro.Parent = parent
+	})
+
+	urn, state, err := constructF(pulumiCtx, req.GetType(), req.GetName(), inputs, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure all outstanding RPCs have completed before proceeding. Also, prevent any new RPCs from happening.
+	pulumiCtx.waitForRPCs()
+	if pulumiCtx.rpcError != nil {
+		return nil, errors.Wrap(pulumiCtx.rpcError, "waiting for RPCs")
+	}
+
+	rpcURN, _, _, err := urn.ToURNOutput().awaitURN(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Serialize all state properties, first by awaiting them, and then marshaling them to the requisite gRPC values.
+	resolvedProps, propertyDeps, _, err := marshalInputs(state)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshaling properties")
+	}
+
+	// Marshal all properties for the RPC call.
+	keepUnknowns := req.GetDryRun()
+	rpcProps, err := plugin.MarshalProperties(
+		resolvedProps,
+		plugin.MarshalOptions{KeepSecrets: true, KeepUnknowns: keepUnknowns, KeepResources: pulumiCtx.keepResources})
+	if err != nil {
+		return nil, errors.Wrap(err, "marshaling properties")
+	}
+
+	// Convert the property dependencies map for RPC and remove duplicates.
+	rpcPropertyDeps := make(map[string]*pulumirpc.ConstructResponse_PropertyDependencies)
+	for k, deps := range propertyDeps {
+		sort.Slice(deps, func(i, j int) bool { return deps[i] < deps[j] })
+
+		urns := make([]string, 0, len(deps))
+		for i, d := range deps {
+			if i > 0 && urns[i-1] == string(d) {
+				continue
+			}
+			urns = append(urns, string(d))
+		}
+
+		rpcPropertyDeps[k] = &pulumirpc.ConstructResponse_PropertyDependencies{
+			Urns: urns,
+		}
+	}
+
+	return &pulumirpc.ConstructResponse{
+		Urn:               string(rpcURN),
+		State:             rpcProps,
+		StateDependencies: rpcPropertyDeps,
+	}, nil
+}
+
+type constructInput struct {
+	value  interface{}
+	secret bool
+	deps   []Resource
+}
+
+// constructInputsMap returns the inputs as a Map.
+func constructInputsMap(inputs map[string]interface{}) Map {
+	result := make(Map, len(inputs))
+	for k, v := range inputs {
+		val := v.(*constructInput)
+		output := newOutput(anyOutputType, val.deps...)
+		output.getState().resolve(val.value, true /*known*/, val.secret, nil)
+		result[k] = output
+	}
+	return result
+}
+
+// constructInputsSetArgs sets the inputs on the given args struct.
+func constructInputsSetArgs(inputs map[string]interface{}, args interface{}) error {
```

Comment:
Note: the implementation of this function isn't quite right, so it's not super important to review the details of it for this PR. I've reimplemented it for prompt values (separate PR on the way), and as part of that, it's leveraging `rpc.unmarshalOutput` to unmarshal the value.